Documentation/manpages/initrd-ls.1.md
initrd-ls(1)

# NAME

initrd-ls - shows contents of initrd archive

# SYNOPSIS

*initrd-ls* [options] initramfs

# DESCRIPTION

Displays initramfs contents in a format similar to ls command. If initramfs
contains more than one cpio archive, utility will show all of them. Archives
can be compressed. In this case, utility will take a look inside.

# OPTIONS

*--no-mtime*
	Hide modification time.

*-b, --brief*
	Show only brief information about archive parts.

*-n, --name*
	Show only filenames.

*-C, --compression*
	Show compression method for each archive.

*-V, --version*
	Show version of program and exit.

*-h, --help*
	Show this text and exit.

# AUTHOR

Written by Alexey Gladkov.

# BUGS

Report bugs to the authors.


NEWS.md
Version 2.39.0

- Feature plymouth:
  + Show splash in synchronous mode to avoid race with plymouth users.
- Feature fsck:
  + Fix typo in boot variable and fix the FASTBOOT parameter.
- Utilities:
  + initrd-scanmod: Add support for kmod > v30-25-g5c004af.

Version 2.38.0

- Runtime:
  + Rename monotonic-timestamp -> timestamp.
  + Use boottime instead of monotonic time.
- Feature dropbear:
  + Mount devpts if not mounted.
- Guess subsystem:
  + Add device suppliers to initrd.
- Remove unsupported bootloader feature.

Version 2.37.0

- Runtime:
  + Reduce few timeouts.
  + Get rid of localdev polling.
  + Convert rootdelay to service.
  + Drop extender scripts.
- Feature zfs:
  + Replace extender script by service.
  + Drop live-mode.

Version 2.36.0

- Runtime:
  + Reimplement ueventd, replace polld by uevent queue.
  + Create a shared libinitramfs library with shared code.
  + Use flock-based locking.
  + Rewrite console locking.
  + Add /dev to fstab and mount it as other filesystems.
  + Do not fail if required mountpoint does not exist.
- New feature:
  + Add debug-procacct. The feature is designed to debug the boot process inside
    the initramfs.
- Feature mdadm:
  + Boot from IMSM.
  + Put mdmon if external metadata detected.
- Feature system-glibc:
  + Do not hardcode min_uid/min_gid values.
- Utilities:
  + udev-rules: Add more udev rule checks.

Version 2.35.0

- Guess subsystem:
  + Do not fail if device filesystem type is swap.
- Misc:
  + Add udev rules parser. This is a new parser that allows you to search for
    syntactic and logical errors. It also allows you to more accurately obtain
    information from the rules.

Version 2.34.0

- Feature guestfs:
  + Add programs used by libguestfs.
  + Add raid modules and udev rules.
  + Config guestfs.mk.example: remove features already required by guestfs.
- Feature kickstart:
  + Do not show rsync progress on serial console.
  + Ask mdadm to create device nodes in /dev.
- Misc:
  + Update busybox 1.35.0.

Version 2.33.0

- Runtime:
  + Use pre/post handlers to extend cmdline service and remove service-specific
    handlers.
  + Replace mountsrc service by post-service handler.
  + Show block/net udev events if rdlog=console.
  + Add timestamps to polld messages.
- Feature zfs:
  + Add OpenZFS feature.
- Feature kickstart:
  + Add zfs support.
  + Add documentattion about zfs commands.
  + Do not try to get UUID for non-device mounts.
- Feature sshfsroot:
  + Add documentattion about root=/dev/sshfs.
- Features mdadm:
  + Use own copy of 63-md-raid-arrays.rules to fix regression with fedora boot
    on raid.
- Misc:
  + Sync gen_init_cpio with upstream.
  + Add documentation about module pattern sets.
  + Add documentation about 9P as root filesystem.

Version 2.32.1

- mk: do not expand functions out of '$(call ...)' context.
- testing: Add cmdline parameter to do not panic.

Version 2.32.0

- Runtime:
  + Reduce rootdelay period if all mountpoints are done, but init program is
    missing.
  + Show proper message if INIT not found.
- Feature luks:
  + Do not overwrite LUKS_CRYPTTAB.

Version 2.31.0

- Runtime:
  + Check more carefully for the presence of the INIT= inside new root partition.
- Feature kickstart:
  + Start all luks after partitioning.
  + Add simple reqpart that automatically creates partitions required by your
    hardware platform.
  + Add part --fstype=efi to create EFI partition with custom mountpoint.
  + Add --hibernation option to part/logvol/raid commands.  This option can be
    used to automatically determine the size of the swap partition big enough
    for hibernation.
  + Add support for fat/vfat filesystem.
  + Add support for zstd/lz4-compressed tarballs in liveimg.
- Feature qemu:
  + Try to add e1000e module.
- Feature network:
  + Fix permissions of /etc/resolv.conf
- Misc:
  + Relax check of /usr. This will allow to migrate the filesystem to /usr.

Version 2.30.0

- Feature kickstart:
  + Close luks partition if we need to change partition table.
  + Add information about what command is being executed.
- Feature rootfs:
  + Create fstab more carefully.
- Feature ucode:
  + Change path in the archive.
- Feature multipath:
  + Add more rules and utils for FC multipath.
- Drop buildinfo feature.
- Misc:
  + Detect separate /usr partition (merged-usr).
  + Generate wiki from Documentation.

Version 2.29.0

- Runtime:
  +  Remove shell service.
- Feature pipeline:
  + Give resume a chance to run.
  + Add wait-resume step.

Version 2.28.0

- Feature lkrg:
  + Support for LKRG v0.9.3-43-g49a3117.
- Feature pipeline:
  + Remove standalone daemon.
- Runtime:
  + Make log messages more readable.
  + Wait until the resume is processed.
- Misc:
  + Add more integration tests.

Version 2.27.2

- Feature luks:
  + Get rid of LUKS_ALL_CRYPTTAB. The LUKS_ALL_CRYPTTAB variable was needed
    until the autodetect worked.

Version 2.27.1

- Feature luks:
  + Fix typo and use keyfile if exists.
  + Add more documentation.

Version 2.27.0

- Feature luks:
  + Add crypttab support.
  + Try to remember the uuid of the luks device and automatically add it to crypttab.
- Feature kickstart:
  + Use /proc/devices to detect sd and virtblk block devices.
- Feature locales:
  + Read system-wide locales.
- Utilities:
  + initrd-ls: Fix infinite loop when unpacking zstd.
  + initrd-put: Add option to exclude files by pattern.
  + initrd-scanmod: module must satisfy all the rules from the ruleset.
  + mkinitrd-make-initrd: Fixed misprint, make-initrd can be found now.
  + make-initrd: Enforce absolute path in TMPDIR.
- Misc:
  + Сheck bzip2 library as the last chance for detection if there is no
    bzip2.pc.

Version 2.26.0

- Runtime:
  + Show a message if the root is not found for more than 15 seconds.
- New fearure:
  + Add locales feature. The feature adds locales and translations of utilities.
- Feature ucode:
  + Don't throw an error if no cpu updates are found for intel.
  + Fix firmware definition for amd cpu.

utils/initrd-ls/initrd-ls.c
#define _GNU_SOURCE
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <getopt.h>
#include <errno.h>
#include <err.h>
#include <libgen.h>
#include <sys/mman.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <fcntl.h>

#include "initrd-common.h"
#include "initrd-cpio.h"
#include "initrd-decompress.h"
#include "initrd-parse.h"
#include "initrd-ls.h"
#include "config.h"

int opts = 0;

static const char cmdopts_s[]        = "bnCVh";
static const struct option cmdopts[] = {
	{ "brief", no_argument, 0, 'b' },
	{ "name", no_argument, 0, 'n' },
	{ "no-mtime", no_argument, 0, 3 },
	{ "compression", no_argument, 0, 'C' },
	{ "version", no_argument, 0, 'V' },
	{ "help", no_argument, 0, 'h' },
	{ NULL, 0, 0, 0 }
};

static void __attribute__((noreturn))
print_help(const char *progname)
{
	printf("Usage: %s [options] initramfs\n"
	       "\n"
	       "Displays initramfs contents in a format similar to ls command.\n"
	       "If initramfs contains more than one cpio archive, utility will show all of them.\n"
	       "Archives can be compressed. In this case, utility will take a look inside.\n"
	       "\n"
	       "Options:\n"
	       "   --no-mtime          Hide modification time;\n"
	       "   -b, --brief         Show only brief information about archive parts;\n"
	       "   -n, --name          Show only filenames;\n"
	       "   -C, --compression   Show compression method for each archive;\n"
	       "   -V, --version       Show version of program and exit;\n"
	       "   -h, --help          Show this text and exit.\n"
	       "\n",
	       progname);
	exit(EXIT_SUCCESS);
}

static void __attribute__((noreturn))
print_version(const char *progname)
{
	printf("%s version " PACKAGE_VERSION "\n"
	       "Written by Alexey Gladkov <gladkov.alexey@gmail.com>\n"
	       "\n"
	       "Copyright (C) 2016-2021  Alexey Gladkov <gladkov.alexey@gmail.com>\n"
	       "This is free software; see the source for copying conditions.  There is NO\n"
	       "warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.\n"
	       "\n",
	       progname);
	exit(EXIT_SUCCESS);
}

int
main(int argc, char **argv)
{
	int c, fd;
	struct stat st;

	while ((c = getopt_long(argc, argv, cmdopts_s, cmdopts, NULL)) != -1) {
		switch (c) {
			case 3:
				opts ^= SHOW_NO_MTIME;
				break;
			case 'b':
				opts ^= SHOW_BRIEF;
				break;
			case 'n':
				opts ^= SHOW_NAME_ONLY;
				break;
			case 'C':
				opts ^= SHOW_COMPRESSION;
				break;
			case 'V':
				print_version(basename(argv[0]));
			case 'h':
				print_help(basename(argv[0]));
			default:
				exit(EXIT_FAILURE);
		}
	}

	if (optind >= argc)
		errx(EXIT_FAILURE, "ERROR: Missing initrd file");

	errno = 0;
	if (stat(argv[optind], &st) == -1) {
		if (errno == ENOENT)
			errx(EXIT_FAILURE, "ERROR: initrd does not exist: %s", argv[optind]);
		err(EXIT_FAILURE, "ERROR: stat: %s", argv[optind]);
	}

	if ((fd = open(argv[optind], O_RDONLY)) == -1)
		err(EXIT_FAILURE, "ERROR: open: %s", argv[optind]);

	unsigned char *addr = mmap(NULL, (size_t) st.st_size, PROT_READ, MAP_PRIVATE | MAP_POPULATE, fd, 0);

	if (addr == MAP_FAILED)
		err(EXIT_FAILURE, "ERROR: mmap");

	unsigned long n_cpio = 0;
	struct stream *s;
	struct list_tail *l, *h;
	struct result res;

	res.streams = NULL;
	res.cpios   = NULL;

	l = list_append(&res.streams, sizeof(struct stream));
	if (l == NULL)
		err(EXIT_FAILURE, "unable to add element to list");
	s = l->data;

	s->addr      = addr;
	s->size      = (unsigned long) st.st_size;
	s->allocated = 0;

	read_stream("raw", s, &res);

	int bytes;
	int max_compress_name = 3;
	char *fmt             = NULL;
	struct cpio *part     = NULL;

	l = res.cpios;
	while (l) {
		part = (struct cpio *) l->data;

		if ((opts & SHOW_BRIEF) || part->type != CPIO_ARCHIVE) {
			l = l->next;
			n_cpio++;
			continue;
		}
		if (opts & SHOW_COMPRESSION) {
			bytes = snprintf(NULL, 0, "%s", part->compress);
			if (bytes > max_compress_name)
				max_compress_name = bytes;
		}
		h = part->headers;
		while (h) {
			preformat(h->data);
			h = h->next;
		}
		l = l->next;
		n_cpio++;
	}

	bytes = snprintf(NULL, 0, "%ju", n_cpio);

	c = (opts & SHOW_COMPRESSION)
	    ? asprintf(&fmt, "%%%dd %%%ds ", bytes, max_compress_name)
	    : asprintf(&fmt, "%%%dd ", bytes);

	if (c == -1)
		err(EXIT_FAILURE, "ERROR: asprintf");

	c = 1;
	l = res.cpios;
	while (l) {
		part = (struct cpio *) l->data;

		if (opts & SHOW_BRIEF) {
			fprintf(stdout, "%d\t", c);

			switch (part->type) {
				case CPIO_ARCHIVE:
					if (strcmp("raw", part->compress))
						fprintf(stdout, "%s compressed ", part->compress);
					fprintf(stdout, "cpio archive");
					break;
				case CPIO_BOOTCONFIG:
					fprintf(stdout, "bootconfig");
					break;
				case CPIO_UNKNOWN:
					fprintf(stdout, "unknown");
					break;
			}

			fprintf(stdout, ", size %ld bytes\n", part->size);

			l = l->next;
			c++;
			continue;
		}
		if (part->type != CPIO_ARCHIVE) {
			l = l->next;
			n_cpio++;
			continue;
		}
		h = part->headers;
		while (h) {
			(opts & SHOW_COMPRESSION)
			? fprintf(stdout, fmt, c, part->compress)
			: fprintf(stdout, fmt, c);
			(opts & SHOW_NAME_ONLY)
			? fprintf(stdout, "%s\n", ((struct cpio_header *) h->data)->name)
			: show_header(h->data);
			h = h->next;
		}
		l = l->next;
		c++;
	}

	free(fmt);

	free_cpios(res.cpios);
	free_streams(res.streams);

	munmap(addr, (size_t) st.st_size);

	return EXIT_SUCCESS;
}
